#517156 [C#テキスト]3章演習問題

sugisaka-being opened the pull request:
3章演習問題の解答を作成しました。

sugisaka-being committed:
#517156 [C#テキスト]3章演習問題

3章演習問題の解答を作成しました。

kurosu-being commented on Chapter3/Chapter3-1-1/Program3-1-1.cs:
```diff
@@ -0,0 +1,47 @@
+            }
+
+            // 4.
+            var wNumbers2 = wNumbers.Select(n => n * 2).ToList();
```
この`wNumbers2 `ですが、
変数は、コード内で見たとき、それが何を指すのかわかる命名にしましょう。
簡潔で明確な変数名を付けることは、コードの可読性において重要です。
例えば
`wNumbers2 `→`wTwiceNumbers`
などというように変更してください。

sugisaka-being replied:
ありがとうございます。
`wNumbers2`から`wTwiceNumbers`へと変数名を修正いたしました。

kurosu-being commented on Chapter3/Chapter3-1-1/Program3-1-1.cs:
```diff
@@ -0,0 +1,47 @@
+
+            // 4.
+            var wNumbers2 = wNumbers.Select(n => n * 2).ToList();
+            foreach (int wN in wNumbers2) {
```
このwNも同様に、適切な命名に変更しましょう。
foreachの場合は、()の中身が`単数名　in 　複数名`　になっていると読みやすい為、
セオリーとして覚えておきましょう！

sugisaka-being replied:
ありがとうございます。
`wN`から`wTwiceNumber`へと変数名を修正いたしました。

kurosu-being commented on Chapter3/Chapter3-1-2/Program3-1-2.cs:
```diff
@@ -0,0 +1,58 @@
+            // 1.
+            Console.WriteLine("都市名を入力してください。");
+            var wLine = Console.ReadLine();
+            Console.WriteLine($"{wLine}は{wCityNames.FindIndex(n => n == wLine)}番目に格納されている。");
```
都市が見つからなかった場合
```
XXXX
は-1番目に格納されている
```
という結果になります。
都市が見つからなかった場合には、問題通り`-1`のみを出力するようにするか、見つからなかった旨をメッセージとして出力するようにしましょう。

sugisaka-being replied:
ありがとうございます。
都市が見つからなかった場合に`-1`のみを出力するように、
`Console.WriteLine($"{wLine}は{wCityNames.FindIndex(n => n == wLine)}番目に格納されている。");`を
`Console.WriteLine(wCityNames.FindIndex(n => n == wLine) != -1 ? $"{wLine}は{wCityNames.FindIndex(n => n == wLine)}番目に格納されている。" : "-1");`に修正いたしました。

kurosu-being replied:
細かいですが、1点だけ！
（修正はこの箇所のみの為、報告は不要です）
`FindIndex`で分岐する場合の条件は wIndex != -1 ではなく、wIndex >= 0 が好ましいです。
理由は、基本的にインデックスを扱うときは 0 以上が存在するを示し、０未満がないことを示します。
ないときは -1 を返す仕様ですが、本質的には -2 でも -100 でもよいです。
インデックスでないを示すのは 0 未満が基本的な条件の扱い方になります。

sugisaka-being replied:
コメントありがとうございます。
`Console.WriteLine(wCityNames.FindIndex(n => n == wLine) >= 0 ? $"{wLine}は{wCityNames.FindIndex(n => n == wLine)}番目に格納されている。" : "-1");`
にコードを修正いたしました・

kurosu-being commented on Chapter3/Chapter3-1-2/Program3-1-2.cs:
```diff
@@ -0,0 +1,58 @@
+            // 3.
+            Console.WriteLine("'o'が含まれている都市名を以下に記述する。");
+            var wCityNames_o = wCityNames.Where(n => n.Contains('o')).ToArray();
+            foreach (var wCityName in wCityNames_o) {
```
`wCityNames_o`ですが、Gaia Cloudでは、変数名であまりアンダーバーを使用していません。
コード内の名前の統一感も可読性向上につながる為、アンダーバーを使用しない形で命名しなおしてみましょう。
同様に、`wCityName `においても、命名しなおしてみてください。

関数においては、コーディング規約で基本的にアンダーバーを使用しないことが書かれています。→[リンク](https://dev-redmine.being.group/projects/sirius-client/wiki/%E3%82%B3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0%E8%A6%8F%E7%B4%84#%E9%96%A2%E6%95%B0%E3%81%AE%E5%91%BD%E5%90%8D%E8%A6%8F%E7%B4%84)

sugisaka-being replied:
ありがとうございます。
`wCityNames_o`を`wCityNamesContainO`
`wCityNames_B`を`wCityNamesStartB`へと変数名を修正いたしました。

kurosu-being replied:
修正ありがとうございます！
[修正不要]
変数名を見たとき、その変数が単数形か複数形なのか末尾ですぐに判断できるような命名だとよりGoodです！
（その変数が配列なのか、ただの変数なのかの推測が容易になる為です。）
例：` ~Name`,  `~Names`

kurosu-being commented on Chapter3/Chapter3-1-2/Program3-1-2.cs:
```diff
@@ -0,0 +1,58 @@
+
+            // 4.
+            Console.WriteLine("'B'で始まる都市名の文字数を以下に記述する。");
+            var wCityNames_B = wCityNames.Where(n => n[0] == 'B').Select(n => n.Length).ToArray();
```
文字列の先頭の文字を取得するのに、LINQをつかうこともできます！
書き換えてみてください。

sugisaka-being replied:
`var wCityNames_B = wCityNames.Where(n => n[0] == 'B').Select(n => n.Length).ToArray();`を
`var wCityNamesStartB = wCityNames.Where(n => n.StartsWith("B")).Select(n => n.Length).ToArray();`へと修正いたしました。

kurosu-being replied:
OKです！
[修正不要]
`StartsWith`など、戻り値が`bool`のメソッドやLINQを使うとき、条件を通るときと通らないときの処理はセットで考えておくと、より保守的な設計につながります！
今回の場合、Bで始まる都市名がなかった場合、特に処理を行っていない為、メッセージが空白になります。このとき、メッセージを出すなどの処理があれば、より明確に条件を通らなかったときの処理がおこなえます。
次回から活かしてみてください💪

sugisaka-being replied:
コメントありがとうございます。
当てはまるものがなかった場合の処理に関しても、
次回からは意識してコードを記述していきたいと思います！
勉強になります！

kurosu-being reviewed (commented):
レビューしました。
軽微なものが多いですが、ご確認お願いします。
また、追加問題も解いてみてください💪

kurosu-being commented:
[追加問題]
Chapter3-1-2の問4と同様に、'n' で終わる都市名を抽出し、アルファベット順にならべたものを出力してください。

m-hotta-being commented on Chapter3/Chapter3-1-1/Program3-1-1.cs:
```diff
@@ -0,0 +1,47 @@
+            // 2.
+            wNumbers.ForEach(n => Console.WriteLine(n / 2.0));
+
+            // 3.
+            var wNumbers1 = wNumbers.Where(n => n >= 50);
+            foreach (var wN in wNumbers1) {
+                Console.WriteLine(wN);
+            }
```
【FYI: 情報共有。アクション不要です。】
こういう書き方も。
```suggestion
            // 3.
            foreach (var wN in wNumbers.Where(n => n >= 50)) {
                Console.WriteLine(wN);
            }
```

こうすることで変数の定義を減らせる。
→変数に名前を付ける回数を減らせる。
→変数に付ける名前で困ることが減る。

sugisaka-being replied:
@kurosu-being 
追加問題の解答を作成いたしました。
```
Console.WriteLine("'n'で終わる都市名をアルファベット順に並べて以下に記述する。。");
var wCityNamesEndN = wCityNames.Where(n => n.EndsWith("n")).OrderBy(n => n).ToArray();
foreach (var wCityNameEndN in wCityNamesEndN) {
    Console.WriteLine(wCityNameEndN);
}
```

sugisaka-being replied:
@m-hotta-being 
コメントありがとうございます。
早速活用させていただきました。コードがすっきりして見やすくなりました。

sugisaka-being committed:
#517156 [C#テキスト]3章演習問題

変数名を変更いたしました。
問題3-1-2の1.において、都市が見つからなかった場合の処理を追加いたしました。
問題3-1-2の4.において、文字列の先頭の文字を取得するためにLINQを用いました。
追加問題の解答を作成しました。
変数の定義を減らして記述しました。

sugisaka-being committed:
#517156 [C#テキスト]3章演習問題

FindIndexで分岐する場合の条件について、書き方を修正いたしました。
